guide_drafts/extending-diesel.md

```diff
@@ -0,0 +1,385 @@
+Diesel doesn't support window functions, or selecting from a subselect.
+Even if Diesel's query builder supported those things,
+this is a case that is easier to reason about in terms of the SQL we want to
+generate.
+
+Let's look at how we would go about adding a `paginate` method to Diesel's query
+builder, to generate that query.
+Let's assume for the time being that we have a struct `Paginated<T>` already.
+We'll look at the specifics of this struct shortly.
+
+If you are creating a struct where you want to manually define the SQL,
+you will need to implement a trait called `QueryFragment`.
+The implementation will look like this:
+
+```
+impl<T> QueryFragment<Pg> for Paginated<T>
+where
+    T: QueryFragment<Pg>,
+{
+    fn walk_ast(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        out.push_sql("SELECT *, COUNT(*) OVER () FROM (");
+        self.query.walk_ast(out.reborrow())?;
+        out.push_sql(") LIMIT ");
+        out.push_bind_param::<BigInt, _>(&self.limit())?;
+        out.push_sql(" OFFSET ");
+        out.push_bind_param::<BigInt, _>(&self.offset())?;
+        Ok(())
+    }
+}
+```
+
+For details on what each method does,
+see the documentation for [`AstPass`].
+One important question to ask whenever you implement `QueryBuilder`
+is whether you are generating a query that is safe to cache.
+The way to answer this question is by asking
+"does this struct generate an unlimited number of potential SQL queries"?
+Typically that is only the case if the body of `walk_ast` contains a for loop.
+If your query is not safe to cache, you *must* call
+`out.unsafe_to_cache_prepared`.
+
+Whenever you implement `QueryFragment`, you also need to implement `QueryId`.
+We can use the [`impl_query_id!`] macro for this.
+Since this struct represents a full query which can be executed,
+we will implement [`RunQueryDsl`] which adds methods like [`execute`] and [`load`].
+Since this query has a return type,
+we'll implement [`Query`] which states the return type as well.
+
+[`AstPass`]: //docs.diesel.rs/diesel/query_builder/struct.AstPass.html
+[`impl_query_id!`]: //docs.diesel.rs/diesel/macro.impl_query_id.html
+[`RunQueryDsl`]: //docs.diesel.rs/diesel/query_dsl/trait.RunQueryDsl.html
+[`execute`]: //docs.diesel.rs/diesel/query_dsl/trait.RunQueryDsl.html#method.execute
+[`load`]: //docs.diesel.rs/diesel/query_dsl/trait.RunQueryDsl.html#method.load
+[`Query`]: //docs.diesel.rs/diesel/query_builder/trait.Query.html
+
+```
+impl_query_id!(Paginated<T>);
+
+impl<T: Query> Query for Paginated<T> {
+    type SqlType = (T::SqlType, BigInt);
+}
+
+impl<T> RunQueryDsl<PgConnection> for Paginated<T> {}
+```
+
+Now that we've implemented all of these things,
+let's look at how we would go about constructing this.
+We'll want to add a `paginate` method to all Diesel queries,
+which specifies which page we're on,
+as well as a `per_page` method which specifies the number of elements per page.
+
+Once again, we can extend existing types with a trait.
```

Comment:
once again? in this guide didn't we didn't write any new traits to extend existing types IIRC

Reply:
We do in section 3 which used to come before this >_<